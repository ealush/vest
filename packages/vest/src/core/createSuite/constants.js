export const SUITE_INIT_ERROR = 'Suite initialization error.';
