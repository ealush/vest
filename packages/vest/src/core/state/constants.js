export const KEY_STATE = '$vestState';
export const KEY_SUITES = '_suites';
export const KEY_CANCELED = '_canceled';
