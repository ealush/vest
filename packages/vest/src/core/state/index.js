import context from '../context';

const SYMBOL_ADD_TO_STATE = Symbol();

export default (function createState() {
  function registerHandler(initialValue) {
    let key;
    function use(patcher) {
      const { stateRef } = context.use();

      if (typeof patcher === 'function') {
        update(patcher);
      }

      function update(patcher) {
        const { stateRef } = context.use();
        const currentState = stateRef.current();
        const prevState = stateRef.prev();

        stateRef.set(
          key,
          typeof patcher === 'function'
            ? patcher(currentState?.[key], prevState?.[key])
            : patcher
        );
      }

      return [stateRef.current()[key], update];
    }

    use[SYMBOL_ADD_TO_STATE] = function (stateKey, ...args) {
      const { stateRef } = context.use();
      key = stateKey;

      if (!Object.prototype.hasOwnProperty.call(stateRef.current(), key)) {
        stateRef.set(
          key,
          typeof initialValue === 'function'
            ? initialValue(...args)
            : initialValue
        );
      }
    };

    return use;
  }

  function createRef(handlers = {}) {
    const state = [];

    function current() {
      return state[0];
    }

    function prev() {
      return state[1];
    }

    function set(key, value) {
      state[0] = {
        ...state[0],
        [key]: value,
      };
    }

    function reset() {
      state.length = 0;

      unshift();
    }

    function unshift() {
      context.run({ stateRef }, () => {
        state.unshift({});
        for (const key in handlers) {
          if (!Object.prototype.hasOwnProperty.call(current(), key)) {
            if (Array.isArray(handlers[key])) {
              /*
                state.createRef({
                  useSuiteId: [useSuiteId, [id, name]],
                })
              */
              handlers[key][0][SYMBOL_ADD_TO_STATE](key, ...handlers[key][1]);
            } else {
              /*
                state.createRef({
                  useSuiteId,
                });
              */
              handlers[key][SYMBOL_ADD_TO_STATE](key);
            }
          }
        }
      });
    }

    const stateRef = {
      current,
      prev,
      reset,
      set,
      unshift,
    };

    context.run({ stateRef }, reset);

    return stateRef;
  }

  return {
    registerHandler,
    createRef,
  };
})();
