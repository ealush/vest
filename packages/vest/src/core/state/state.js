import callEach from 'callEach';
import context from 'ctx';
import isFunction from 'isFunction';

export default (function createState() {
  function registerHandler(initialValue) {
    let key;
    function use(patcher) {
      const { stateRef, reg } = context.use();

      // Register state handler
      if (reg) {
        key = reg.key;
        if (!stateRef.current().hasOwnProperty(key)) {
          stateRef.set(
            key,
            isFunction(initialValue)
              ? initialValue.apply(null, reg.args)
              : initialValue
          );
        }
        return;
      }

      if (isFunction(patcher)) {
        update(patcher);
      }

      function update(patcher) {
        const { stateRef } = context.use();
        const currentState = stateRef.current();

        stateRef.set(
          key,
          isFunction(patcher) ? patcher(currentState[key]) : patcher
        );
      }

      return [stateRef.current()[key], update];
    }

    return use;
  }

  function createRef(handlers = {}) {
    let state = {};
    const registeredHandlers = [];

    function current() {
      return state;
    }

    function set(key, value) {
      Object.assign(state, { [key]: value });
    }

    function reset() {
      state.length = 0;

      state = {};
      callEach(registeredHandlers);
    }

    const stateRef = {
      current,
      reset,
      set,
    };

    for (const key in handlers) {
      const [handler, args] = [].concat(handlers[key]);
      registeredHandlers.push(
        context.bind({ stateRef, reg: { key, args } }, handler)
      );
    }

    context.run({ stateRef }, reset);

    return stateRef;
  }

  return {
    registerHandler,
    createRef,
  };
})();
