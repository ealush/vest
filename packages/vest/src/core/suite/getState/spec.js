import * as state from '..';
import runRegisterSuite from '../../../../testUtils/runRegisterSuite';
import getSuiteState from '.';

const suiteName = 'suite_1';
const suiteId = 'suiteId_1';
const suiteId_2 = 'suiteId_3';

describe('getSuiteState', () => {
  it('Should return current suite state by key', () => {
    runRegisterSuite({ name: suiteName, suiteId });
    runRegisterSuite({ name: suiteName, suiteId: suiteId_2 });
    const [currentState] = state.getSuite(suiteId);
    expect(getSuiteState(suiteId)).toBe(currentState);
  });
});
