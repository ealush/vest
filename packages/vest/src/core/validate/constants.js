/**
 * @type {String}
 */
export const SUITE_INIT_ERROR = "Suite initialization error.";
