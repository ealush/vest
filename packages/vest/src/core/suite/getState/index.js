import { getSuite } from '../../state';

/**
 * Retrieves most recent suite state.
 * @param {string} suiteId
 * @returns {Object} Current suite state.
 */
const getSuiteState = suiteId => getSuite(suiteId)[0];

export default getSuiteState;
