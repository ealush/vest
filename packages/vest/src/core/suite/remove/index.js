import { setSuites } from '..';
import throwError from '../../../lib/throwError';
import { setCanceled } from '../../test/lib/canceled';
import getState from '../getState';

/**
 * Cleans up a suite from state.
 * @param {string} suiteId
 */
const remove = suiteId => {
  if (!suiteId) {
    throwError('`vest.remove` must be called with suiteId.');
  }

  const suite = getState(suiteId);
  if (!suite) {
    return;
  }

  setCanceled(...(suite.pending || []));
  setCanceled(...(suite.lagging || []));

  setSuites(state => {
    delete state[suiteId];
    return state;
  });
};

export default remove;
