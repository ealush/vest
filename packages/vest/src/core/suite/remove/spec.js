import _ from 'lodash';
import vest from '../../..';
import resetState from '../../../../testUtils/resetState';
import runRegister from '../../../../testUtils/runRegister';
import testDummy from '../../../../testUtils/testDummy';
import * as state from '../../state';
import { KEY_CANCELED, KEY_SUITES } from '../../state/constants';
import getState from '../getState';
import remove from '.';

const suiteId = 'suite_id';
let currentState;

describe('remove', () => {
  const createSuite = () =>
    vest.create(suiteId, skip => {
      vest.skip(skip);

      testDummy(vest).failingAsync('field_1', { time: 250 });

      testDummy(vest).failingAsync('field_2', { time: 250 });

      testDummy(vest).failingAsync('field_3', { time: 250 });

      testDummy(vest).failingAsync('field_4', { time: 250 });
    });

  beforeAll(() => {
    resetState();
  });
  describe('When invoked without suiteId', () => {
    it('Should throw', () => {
      expect(() => remove()).toThrow(
        '`vest.remove` must be called with suiteId'
      );
    });
  });

  describe('When suite does not exist in state', () => {
    beforeEach(() => {
      runRegister({ name: suiteId, suiteId });
      currentState = _.cloneDeep(state.get());
    });

    it('Should throw an error', () => {
      expect(currentState).toEqual(state.get());
      expect(() => remove('nonexistent_suite')).toThrow(Error);
    });
  });

  describe('When suite exists', () => {
    let validate;

    beforeEach(() => {
      validate = createSuite();
      validate('field_2');
      validate(['field_1', 'field_3']);
    });

    test.skipOnWatch(
      'Sanity - making sure everything works as it should',
      () => {
        const suiteState = getState(suiteId);
        expect(suiteState.lagging).toHaveLength(2);
        expect(suiteState.pending).toHaveLength(2);
        expect(suiteState.lagging[0].fieldName).toBe('field_1');
        expect(suiteState.pending[0].fieldName).toBe('field_2');
        expect(suiteState.lagging[1].fieldName).toBe('field_3');
        expect(suiteState.pending[1].fieldName).toBe('field_4');
        expect(state.get()).toMatchSnapshot();
      }
    );

    it('Should remove suite from state', () => {
      expect(state.get()[KEY_SUITES]).toHaveProperty(suiteId);
      expect(() => getState(suiteId)).not.toThrow();
      remove(suiteId);
      expect(() => getState(suiteId)).toThrow();
      expect(state.get()[KEY_SUITES]).not.toHaveProperty(suiteId);
    });

    it('Should set all pending and lagging tests as canceled', () => {
      const previouslyCanceled = state.get()[KEY_CANCELED];
      const currentState = getState(suiteId);
      const allCanceled = [
        ...currentState.lagging,
        ...currentState.pending,
      ].reduce((canceled, { id }) => Object.assign(canceled, { [id]: true }), {
        ...previouslyCanceled,
      });
      expect(state.get()[KEY_CANCELED]).not.toEqual(allCanceled);
      remove(suiteId);
      expect(state.get()[KEY_CANCELED]).toEqual(allCanceled);
    });
  });
});
