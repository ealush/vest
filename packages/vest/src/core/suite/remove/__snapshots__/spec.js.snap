// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`remove When suite exists Sanity - making sure everything works as it should 1`] = `
Object {
  "_canceled": Object {
    "3": true,
  },
  "_suites": Object {
    "suite_id": Array [
      Object {
        "doneCallbacks": Array [],
        "fieldCallbacks": Object {},
        "groups": Object {},
        "lagging": Array [
          VestTest {
            "asyncTest": Promise {},
            "failed": false,
            "fieldName": "field_1",
            "id": "0",
            "isWarning": false,
            "statement": undefined,
            "suiteId": "suite_id",
            "testFn": [MockFunction] {
              "calls": Array [
                Array [],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": Promise {},
                },
              ],
            },
          },
          VestTest {
            "asyncTest": Promise {},
            "failed": false,
            "fieldName": "field_3",
            "id": "2",
            "isWarning": false,
            "statement": undefined,
            "suiteId": "suite_id",
            "testFn": [MockFunction] {
              "calls": Array [
                Array [],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": Promise {},
                },
              ],
            },
          },
        ],
        "name": "suite_id",
        "pending": Array [
          VestTest {
            "asyncTest": Promise {},
            "failed": false,
            "fieldName": "field_2",
            "id": "5",
            "isWarning": false,
            "statement": undefined,
            "suiteId": "suite_id",
            "testFn": [MockFunction] {
              "calls": Array [
                Array [],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": Promise {},
                },
              ],
            },
          },
          VestTest {
            "asyncTest": Promise {},
            "failed": false,
            "fieldName": "field_4",
            "id": "7",
            "isWarning": false,
            "statement": undefined,
            "suiteId": "suite_id",
            "testFn": [MockFunction] {
              "calls": Array [
                Array [],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": Promise {},
                },
              ],
            },
          },
        ],
        "suiteId": "suite_id",
        "testObjects": Array [
          VestTest {
            "asyncTest": Promise {},
            "failed": false,
            "fieldName": "field_2",
            "id": "5",
            "isWarning": false,
            "statement": undefined,
            "suiteId": "suite_id",
            "testFn": [MockFunction] {
              "calls": Array [
                Array [],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": Promise {},
                },
              ],
            },
          },
          VestTest {
            "asyncTest": Promise {},
            "failed": false,
            "fieldName": "field_4",
            "id": "7",
            "isWarning": false,
            "statement": undefined,
            "suiteId": "suite_id",
            "testFn": [MockFunction] {
              "calls": Array [
                Array [],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": Promise {},
                },
              ],
            },
          },
          VestTest {
            "asyncTest": Promise {},
            "failed": false,
            "fieldName": "field_1",
            "id": "0",
            "isWarning": false,
            "statement": undefined,
            "suiteId": "suite_id",
            "testFn": [MockFunction] {
              "calls": Array [
                Array [],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": Promise {},
                },
              ],
            },
          },
          VestTest {
            "asyncTest": Promise {},
            "failed": false,
            "fieldName": "field_3",
            "id": "2",
            "isWarning": false,
            "statement": undefined,
            "suiteId": "suite_id",
            "testFn": [MockFunction] {
              "calls": Array [
                Array [],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": Promise {},
                },
              ],
            },
          },
        ],
        "tests": Object {},
      },
      Object {
        "doneCallbacks": Array [],
        "fieldCallbacks": Object {},
        "groups": Object {},
        "lagging": null,
        "name": "suite_id",
        "pending": null,
        "suiteId": "suite_id",
        "testObjects": Array [
          VestTest {
            "asyncTest": Promise {},
            "failed": false,
            "fieldName": "field_1",
            "id": "0",
            "isWarning": false,
            "statement": undefined,
            "suiteId": "suite_id",
            "testFn": [MockFunction] {
              "calls": Array [
                Array [],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": Promise {},
                },
              ],
            },
          },
          VestTest {
            "asyncTest": Promise {},
            "failed": false,
            "fieldName": "field_3",
            "id": "2",
            "isWarning": false,
            "statement": undefined,
            "suiteId": "suite_id",
            "testFn": [MockFunction] {
              "calls": Array [
                Array [],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": Promise {},
                },
              ],
            },
          },
          VestTest {
            "asyncTest": Promise {},
            "failed": false,
            "fieldName": "field_4",
            "id": "3",
            "isWarning": false,
            "statement": undefined,
            "suiteId": "suite_id",
            "testFn": [MockFunction] {
              "calls": Array [
                Array [],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": Promise {},
                },
              ],
            },
          },
        ],
        "tests": Object {},
      },
    ],
  },
}
`;
